src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/JCloudsProvider.java

```diff
@@ -215,15 +225,46 @@ protected String buildScriptToExecuteString(InstanceScript instanceScript) {
 
     }
 
-    private RunScriptOptions buildScriptOptions(InstanceScript instanceScript) {
-        logger.info("Script options: userLogin=" + getVmUserLogin());
-        return Optional.ofNullable(instanceScript.getCredentials())
-                       .map(credentials -> RunScriptOptions.Builder.runAsRoot(false)
-                                                                   .overrideLoginCredentials(new LoginCredentials.Builder().user(credentials.getUsername())
-                                                                                                                           .password(credentials.getPassword())
-                                                                                                                           .authenticateSudo(false)
-                                                                                                                           .build()))
-                       .orElse(RunScriptOptions.Builder.overrideLoginUser(getVmUserLogin()));
+    private RunScriptOptions buildScriptOptions(InstanceScript instanceScript, String instanceId,
+            Infrastructure infrastructure, String instanceTag) {
+        return Optional.ofNullable(instanceScript.getCredentials()).map(credentials -> {
+
+            // retrieve the passed username or read it from the property file
+            // only in the case where credentials are provided by the user
+            String username = Optional.ofNullable(credentials.getUsername())
+                                      .filter(StringUtils::isNotEmpty)
+                                      .filter(StringUtils::isNotBlank)
+                                      .orElse(getVmUserLogin());
+            if (credentials.getPrivateKey() != null) {
```

Comment:
This method is now a bit too complex and also it contains specific logic only for AWS. This class is abstract and the design allows specific implementation (You can put specific AWS logic in the AWSEC2JCloudsProvider class)

Reply:
Agree. I have changed the if/else into a template method call.